### DISCUSSION TITLE
The difference between forward and backward inference in lesson 2?

### DISCUSSION THREAD

**Discussion by kbs-code** [microsoft/AI-For-Beginners#288] [Q&A]
After reading about the explanation between forward and backward inference, I still don't get the difference.  I ran all the code in the Animals notebook and it seems that all cells output the same thing but use different methods to get there.  The backward inference cell gives me "tiger" if I answer the right questions and the forward inference cell gives me "tiger" if I input the right facts.  So how is this different?  The only difference in these two approaches is that one is asking me a series of questions and the other isn't.  Can someone maybe abstract this concept into another example?  How is forward/backward inference applied elsewhere?

**Comment by mouseart** [comment]
Understanding the difference between Forward Inference and Backward Inference indeed requires some level of abstract thinking. While in the "Animals notebook" example you mentioned, both methods lead to the same conclusion - "tiger", these two reasoning approaches differ in their operational methods and application contexts.

1.  Forward Inference: This method starts with known facts and rules, and gradually deduces new facts. For example, if we know that "all cats are mammals" and "a tiger is a cat", then we can deduce through forward inference that "a tiger is a mammal". In this method, reasoning is actively "pushing out" conclusions from known information. 
2.  Backward Inference: Conversely, backward inference starts from a hypothesis (or question) we want to prove, and traces back to known facts. For example, if we want to prove "a tiger is a mammal", we start with this hypothesis and then check all the rules and facts that support this hypothesis, such as "a tiger is a cat" and "all cats are mammals". In this method, reasoning is tracing back from the conclusion to the known facts to validate the hypothesis. 

Differences:

●  In the "Animals notebook" example, forward inference derives a conclusion (what animal it is) based on inputted facts (specific characteristics of an animal). Backward inference, however, validates a specific conclusion (is this a tiger?) through answering a series of questions. 
●  Forward inference is akin to building a "fact network" and adding new nodes (facts) to this network, whereas backward inference is more like finding the path to a specific node (conclusion) along this network. 

Other Application Examples:

●  In medical diagnosis, forward inference can be used to infer possible diseases based on a patient's symptoms and known disease patterns; backward inference can be used to validate the likelihood of a specific disease, i.e., starting from the disease and checking if the patient has all related symptoms. 
●  In business decision-making, forward inference can be used to predict future trends based on market data and consumer behavior analysis; backward inference can be used to determine the various conditions needed to achieve a specific business goal. 

I hope these explanations help you better understand the differences between forward and backward reasoning and their applications.
